Tasks/DownloadPackageV1/connections.ts

```diff
@@ -0,0 +1,43 @@
+import { BearerCredentialHandler } from "azure-devops-node-api/handlers/bearertoken";
+import { WebApi } from "azure-devops-node-api";
+import { IRequestOptions } from "azure-devops-node-api/interfaces/common/VsoBaseInterfaces";
+
+import * as tl from 'vsts-task-lib/task';
+import * as locationUtility from "packaging-common/locationUtilities";
+
+export class BearerHandlerForPresignedUrls extends BearerCredentialHandler {
+    prepareRequest(options) {
+        // If we have a presigned blobstore url, don't add auth header
+        if (this.isPreSignedUrl(options)) {
+            delete options.headers["Authorization"];
+            delete options.headers["X-TFS-FedAuthRedirect"];
+        } else {
+            options.headers["Authorization"] = "Bearer " + this.token;
+            options.headers["X-TFS-FedAuthRedirect"] = "Suppress";
+        }
+    }
+
+    isPreSignedUrl(options: any): boolean {
+        return (
+            options.host &&
+            options.host.endsWith("blob.core.windows.net") &&
+            options.path &&
+            options.path.includes("&sig=")
+        );
+    }
+}
+
+export function getConnection(areaId: string, collectionUrl: string): Promise<WebApi> {
+    var accessToken = locationUtility.getSystemAccessToken();
+    return locationUtility
+        .getServiceUriFromAreaId(collectionUrl, accessToken, areaId)
+        .then(url => {
+            const options: IRequestOptions = {
+                proxy: tl.getHttpProxyConfiguration(url)
+            };
+            return new WebApi(url, new BearerHandlerForPresignedUrls(accessToken), options);
```

Comment:
where is this used? I only see it used for connecting to feed & packaging, both of which would be fine with the inbuild bearerhandler.

Reply:
The issue is with redirects to blobstore. Python for now, and the rest of the protocols soon, will send a pre-signed blobstore url redirect when we send a request to the package download url. The regular bearer handler adds the authorization header to the following redirect request to blobstore as well, which causes a 401. This class doesn't add the auth header if the url is blobstore, and pre-signed.